pinot-segment-local/src/main/java/org/apache/pinot/segment/local/realtime/impl/json/MutableJsonIndex.java

```diff
@@ -0,0 +1,295 @@
+import org.apache.pinot.segment.spi.index.reader.JsonIndexReader;
+import org.apache.pinot.spi.exception.BadQueryRequestException;
+import org.apache.pinot.spi.utils.JsonUtils;
+import org.apache.pinot.sql.parsers.CalciteSqlParser;
+import org.roaringbitmap.IntConsumer;
+import org.roaringbitmap.RoaringBitmap;
+import org.roaringbitmap.buffer.MutableRoaringBitmap;
+
+import static org.apache.pinot.common.request.context.FilterContext.Type.PREDICATE;
+
+
+/**
+ * Json index for mutable segment.
+ */
+public class MutableJsonIndex implements JsonIndexReader {
+  private final Map<String, RoaringBitmap> _postingListMap;
+  private final IntList _docIdMapping;
+  private final ReentrantReadWriteLock.ReadLock _readLock;
+  private final ReentrantReadWriteLock.WriteLock _writeLock;
+
+  private int _nextDocId;
+  private int _nextFlattenedDocId;
+
+  public MutableJsonIndex() {
+    _postingListMap = new HashMap<>();
+    _docIdMapping = new IntArrayList();
+
+    ReentrantReadWriteLock readWriteLock = new ReentrantReadWriteLock();
+    _readLock = readWriteLock.readLock();
+    _writeLock = readWriteLock.writeLock();
+  }
+
+  /**
+   * Adds the next json value.
+   */
+  public void add(String jsonString)
+      throws IOException {
+    List<Map<String, String>> flattenedRecords = JsonUtils.flatten(JsonUtils.stringToJsonNode(jsonString));
+    _writeLock.lock();
+    try {
+      addFlattenedRecords(flattenedRecords);
+    } finally {
+      _writeLock.unlock();
+    }
+  }
+
+  /**
+   * Adds the flattened records for the next document.
+   */
+  private void addFlattenedRecords(List<Map<String, String>> records) {
+    int numRecords = records.size();
+    Preconditions
+        .checkState(_nextFlattenedDocId + numRecords >= 0, "Got more than %s flattened records", Integer.MAX_VALUE);
+    for (int i = 0; i < numRecords; i++) {
+      _docIdMapping.add(_nextDocId);
+    }
+    // TODO: Consider storing tuples as the key of the posting list so that the strings can be reused, and the hashcode
+    //       can be cached.
+    for (Map<String, String> record : records) {
+      for (Map.Entry<String, String> entry : record.entrySet()) {
+        // Put both key and key-value into the posting list. Key is useful for checking if a key exists in the json.
+        String key = entry.getKey();
+        _postingListMap.computeIfAbsent(key, k -> new RoaringBitmap()).add(_nextFlattenedDocId);
+        String keyValue = key + JsonIndexCreator.KEY_VALUE_SEPARATOR + entry.getValue();
+        _postingListMap.computeIfAbsent(keyValue, k -> new RoaringBitmap()).add(_nextFlattenedDocId);
+      }
+      _nextFlattenedDocId++;
+    }
+    _nextDocId++;
+  }
+
+  @Override
+  public MutableRoaringBitmap getMatchingDocIds(String filterString) {
+    FilterContext filter;
+    try {
+      filter = RequestContextUtils.getFilter(CalciteSqlParser.compileToExpression(filterString));
+    } catch (Exception e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)